tools/perf_util/src/main/java/com/msft/send_perf_metrics/App.java

```diff
@@ -94,63 +90,86 @@ private static void processPerfMetrics(final List<Path> perf_metrics, String com
 		}
 	}
 
+
 	static private void loadMetricsIntoMySQL(java.sql.Connection conn, String commit_id, String batch_id,
 											 JSONObject json_object) throws Exception {
 
-		try (java.sql.PreparedStatement st = conn.prepareStatement(
-				"INSERT INTO perf_test_training_data (BatchId,CommitId,Model,ModelName,DisplayName,UseMixedPrecision,Optimizer,BatchSize,SeqLen,PredictionsPerSeq," +
-						"NumOfBatches,WeightUpdateSteps,Round,GradAccSteps,AvgTimePerBatch,Throughput,StabilizedThroughput,TotalTime,AvgCPU,Memory,RunConfig,Time) " +
-						"values (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,Now())"
-						+ "  ON DUPLICATE KEY UPDATE AvgTimePerBatch=?,Throughput=?,StabilizedThroughput=?,TotalTime=?,AvgCPU=?,Memory=?")) {
-
-			int i = 0;
-
-			// unique key section
-			st.setString(++i, batch_id);
-			st.setString(++i, commit_id.substring(0, 8));
-			st.setString(++i, (String) json_object.get("Model"));
-			st.setString(++i, (String) json_object.get("ModelName"));
-			st.setString(++i, (String) json_object.get("DisplayName"));
-			st.setBoolean(++i, (Boolean) json_object.get("UseMixedPrecision"));
-			st.setString(++i, (String) json_object.get("Optimizer"));
-			st.setInt(++i, (int)(long) json_object.get("BatchSize"));
-
-			// non-key section
-			JSONObject properties = (JSONObject) json_object.get("DerivedProperties");
-			if (properties != null) {
-				if (properties.get("SeqLen") == null) 				//  mysql allows null value in unique key column
-					st.setNull(++i, Types.INTEGER);
-				else
-					st.setInt(++i, Integer.parseInt((String) properties.get("SeqLen")));
-
-				if (properties.get("PredictionsPerSeq") == null) 				//  mysql allows null value in unique key column
-					st.setNull(++i, Types.INTEGER);
-				else
-					st.setInt(++i, Integer.parseInt((String) properties.get("PredictionsPerSeq")));
+		// field name -> json value
+		Map<String, Object> field_mapping = new HashMap();
+		Set<String> update_on_duplicate_fields =
+			new HashSet<> (Arrays.asList("AvgTimePerBatch", "Throughput", "StabilizedThroughput", "TotalTime", "AvgCPU", "Memory"));
+
+		field_mapping.put("BatchId", batch_id);
+		field_mapping.put("CommitId", commit_id.substring(0, 8));
+		json_object.forEach((key, value) -> {
+			if (key.equals("DerivedProperties")) {
+				JSONObject properties = (JSONObject) json_object.get("DerivedProperties");
+				properties.forEach((sub_key, sub_value) -> {
+					field_mapping.put((String)sub_key, sub_value);
+				});
+			} else if (key.equals("Memory")) {
```

Comment:
should we just produce the data in the correct units? the fewer special cases the better :)

Reply:
I'd like to keep the raw data separated from normalization here, so will leave it as is for now.

Reply:
on the second thought, let me do t hat.